### DISCUSSION TITLE
feat: Add ability to hide / fold some sections of chat

### DISCUSSION THREAD

**Discussion by niba** [olimorris/codecompanion.nvim#60] [Ideas]
Some data that goes to the chat buffer clutters it too much:

- System prompt - It is good to see it when you start your journey or when you're testing new things, but when you're familiar with your prompts, it only takes up space. The system prompt can be really long if someone wants to be very specific.
- Buffer contents - Buffer contents and even large selected text can be really long.

It would be perfect if you could hide or at least fold this data. Copilot chat does that, and other Neovim plugins (like CopilotChat) do as well.
I'm quite new to Neovim, so maybe I'm not familiar with some workarounds on how to achieve that.

Thanks for great plugin!

**Comment by olimorris** [comment]
Can you share some screen shots? By default the chat buffer should automatically scroll to the bottom of the buffer

**Reply by niba** [reply]
It scrolls to the bottom. It's more about UX.

The system prompt is usually not important. When I scroll to the top, I prefer to see the beginning of my chat conversation rather than a multi-line long prompt. I believe that some people also prefer to hide it (for sharing screens, etc.).

I open an AI chat with my loaded buffers (@buffers). The chat automatically scrolls to the bottom, but it's hard to see which buffers were loaded. It would make more sense to display them as a list of paths. I don't need to see the entire content of them. In the future, I would like to add the possibility to select specific buffers to save some money.
The person who requested this feature (@buffers) also mentioned folding functionality: https://github.com/olimorris/codecompanion.nvim/discussions/57

**Reply by olimorris** [reply]
Some fair points. Perhaps there's a case for adding a config option to hide system prompts from the buffer altogether. The large system prompt that comes with the plugin by default is always added prior to sending to the LLM, for example.

I'm currently working through a new feature which adds helpers to the chat buffer itself. So you can type things like `@buffers` to add open and loaded buffers as context (which won't appear as text in the buffer) or `@bufferq` to load buffers from the quickfix window.

So I'll add your ask to the backlog but can't promise it anytime soon.

**Reply by olimorris** [reply]
@niba can you check the latest commit? I added some folding for buffer content. Looks a little messy right now but works

**Reply by niba** [reply]
Thanks, awesome! It looks good. The only suggestion for minimal improvement that I have is to be able to see file_path without file_content (nested folding?).

btw I like the idea to use commands / helpers in chat buffer like @buffer

**Reply by olimorris** [reply]
Great to hear it. I've added a cmp source for them now as well so you just need to type `@` to trigger them appearing in the chat buffer.

Take your point on the seeing the file_path element. I'll have a think about how that could be incorporated. Equally, would be open to a PR on this.

**Comment by olimorris** [comment]
@niba - as per 49559c14b57421240748afe6a95626eac407925c, I'm hiding system prompts by default. The UI is looking much cleaner.

**Reply by niba** [reply]
@olimorris  Thanks for letting me know. It is indeed much cleaner. I'm only wondering that maybe there should be an option to show prompt description (based on description field) like "Senior Developer Typescript" at the top. Something like that

![image](https://github.com/user-attachments/assets/47f9dbdd-13d2-4bc5-98ce-6ecf1cabe08b)

**Reply by olimorris** [reply]
That's a very interesting idea. There's a few more UI enhancements I'm still to make (thinking of adding a keymap for `?` which would display all of the available keymaps in the chat buffer).

I had been thinking of removing the `Senior Dev` prompt from the plugin all together. The reason being, I think the default system prompts are now really strong and the additional context that's provided from the `Senior Dev` prompt doesn't add anything. Keen to get you thoughts

**Reply by niba** [reply]
I mentioned this description because, when you try to find a better prompt, it can be hard to differentiate which output comes from which prompt after some time. If you test prompts that produce similar outputs and save the chats to compare them, it can be difficult to distinguish between them without the description or prompt text.

However, if you have a favorite prompt or just use the default one, you mostly use the same one consistently and don't care about the description or prompt text. That's why I'm not sure if it's worth adding.

I don't know if the `Senior Dev` is better or not because I don't use it. I created my own prompt based on this: https://www.reddit.com/r/ClaudeAI/comments/1dwra38/sonnet_35_for_coding_system_prompt which I use for everything. However, I think fewer choices provide a better UX. When introducing a new person, it would be better to have just 'CodeCompanion Chat' rather than both 'CodeCompanion default chat' and 'CodeCompanion Senior Dev chat.' If the difference is subtle and the default prompts are strong, then I agree with you that removing `Senior Dev` is a good idea
